torchvision/prototype/datasets/utils/_dataset.py

```diff
@@ -0,0 +1,220 @@
+import abc
+import io
+import os
+import pathlib
+import textwrap
+from collections import Mapping
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Union,
+    NoReturn,
+    Iterable,
+    Tuple,
+)
+
+import torch
+from torch.utils.data import IterDataPipe
+
+from torchvision.prototype.datasets.utils._internal import (
+    add_suggestion,
+    sequence_to_str,
+)
+from ._resource import OnlineResource
+
+
+def make_repr(name: str, items: Iterable[Tuple[str, Any]]):
+    def to_str(sep: str) -> str:
+        return sep.join([f"{key}={value}" for key, value in items])
+
+    prefix = f"{name}("
+    postfix = ")"
+    body = to_str(", ")
+
+    line_length = int(os.environ.get("COLUMNS", 80))
+    body_too_long = (len(prefix) + len(body) + len(postfix)) > line_length
+    multiline_body = len(str(body).splitlines()) > 1
+    if not (body_too_long or multiline_body):
+        return prefix + body + postfix
+
+    body = textwrap.indent(to_str(",\n"), " " * 2)
+    return f"{prefix}\n{body}\n{postfix}"
+
+
+class DatasetConfig(Mapping):
+    def __init__(self, *args, **kwargs):
+        data = dict(*args, **kwargs)
+        self.__dict__["__data__"] = data
+        self.__dict__["__final_hash__"] = hash(tuple(data.items()))
+
+    def __getitem__(self, name: str) -> Any:
+        return self.__dict__["__data__"][name]
+
+    def __iter__(self):
+        return iter(self.__dict__["__data__"].keys())
+
+    def __len__(self):
+        return len(self.__dict__["__data__"])
+
+    def __getattr__(self, name: str) -> Any:
+        try:
+            return self[name]
+        except KeyError as error:
+            raise AttributeError(
+                f"'{type(self).__name__}' object has no attribute '{name}'"
+            ) from error
+
+    def __setitem__(self, key: Any, value: Any) -> NoReturn:
+        raise RuntimeError(f"'{type(self).__name__}' object is immutable")
+
+    def __setattr__(self, key: Any, value: Any) -> NoReturn:
+        raise RuntimeError(f"'{type(self).__name__}' object is immutable")
+
+    def __delitem__(self, key: Any) -> NoReturn:
+        raise RuntimeError(f"'{type(self).__name__}' object is immutable")
+
+    def __delattr__(self, item: Any) -> NoReturn:
+        raise RuntimeError(f"'{type(self).__name__}' object is immutable")
+
+    def __hash__(self) -> int:
+        return self.__dict__["__final_hash__"]
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, DatasetConfig):
+            return NotImplemented
+
+        return hash(self) == hash(other)
+
+    def __repr__(self) -> str:
+        return make_repr(type(self).__name__, self.items())
+
+
+class DatasetInfo:
+    def __init__(
+        self,
+        name: str,
+        *,
+        categories: Union[int, Sequence[str], str, pathlib.Path],
```

Comment:
For the future: this assumes that all we care about for datasets are the categories. But sometimes datasets don't have categories (i.e., for depth estimation), or it might have more info that we might want to expose (meta-categories).

If we assume that all datasets will return dicts, we could instead let this return the types for each one of the outputs of the dataset. But we can postpone the discussion to the future